src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala

```diff
@@ -0,0 +1,92 @@
+package loamstream.googlecloud
+
+import loamstream.util.Loggable
+import loamstream.util.Paths.normalize
+import scala.util.Try
+import scala.util.Success
+import loamstream.util.Tries
+import loamstream.util.ExitCodes
+import scala.concurrent.duration.Duration
+import loamstream.util.BashScript
+import java.nio.file.Paths
+
+/**
+ * @author clint
+ * Nov 28, 2016
+ */
+final class CloudSdkDataProcWrapper private[googlecloud] (
+  config: GoogleCloudConfig,
+  runCommand: Seq[String] => Int = CloudSdkDataProcWrapper.runCommand) extends 
+      DataProcClient.CanStop with DataProcClient.CanTellIfRunning with Loggable {
+
+  import CloudSdkDataProcWrapper.{
+    deleteClusterTokens,
+    isClusterRunningTokens
+  }
+
+  override def stopCluster(): Unit = {
+    debug(s"Deleting cluster '${config.clusterId}'")
+
+    val exitCode = runCommand(deleteClusterTokens(config))
+
+    ExitCodes.throwIfFailure(exitCode)
+  }
+
+  override def isClusterRunning: Boolean = {
+    val exitCode = runCommand(isClusterRunningTokens(config))
+
+    val result = ExitCodes.isSuccess(exitCode)
+
+    debug(s"Cluster '${config.clusterId}' running? $result")
+
+    result
+  }
+}
+
+object CloudSdkDataProcWrapper extends Loggable {
+  def fromConfig(config: GoogleCloudConfig): Try[CloudSdkDataProcWrapper] = {
+    val gcloudBinary = config.gcloudBinary.toFile
+
+    if (gcloudBinary.exists && gcloudBinary.canExecute) {
+      Success(new CloudSdkDataProcWrapper(config))
+    } else {
+      Tries.failure(s"gcloud executable not found at ${config.gcloudBinary} or not executable")
+    }
+  }
+
+  private[googlecloud] def deleteClusterTokens(config: GoogleCloudConfig): Seq[String] = {
+    gcloudTokens(config)("delete")(config.clusterId)
```

Comment:
Just wondering... these "tokens" functions look like they are "constant" for the config and only use a `GoogleCloudConfig` in order to produce the them. Why are these functions not part of `GoogleCloudConfig` and maybe even just a (lazy?) `val` member?

Just wondering if there's some rational behind that would impact later code or that I should be aware of for the future.

Reply:
A few things: I liked the `*Config` classes just being dumb containers of data, since that removes the possibility for surprises regarding how they're de-serialized from typesafe-config objects.  I don't know for sure about `Ficus` but it's certainly not rare for macro- and reflection-based libs like that to get confused by "other" members that aren't mapped to anything in the serialized form.

I also wanted to keep the particulars about how command-lines are built from data (or if command lines are built at all) tightly scoped to the places where those command lines are actually needed and used.  `CloudSdkDataProcWrapper` is just one implementation of a "thing that does Google dataproc cluster-management operations" that happens to build and run commands locally.  I didn't want that detail to leak out and get baked into other APIs or couplings.

The methods that make command-line tokens could be vals (lazy or otherwise).  It was just marginally easier/more straightforward to test
```scala
member(knownInput) == expectedOutput
```
instead of 
```scala
WrappingClass(knownInput).member == expectedOutput
```
That's a small thing, and those two forms _are_ functionally equivalent.